File: code/Project_4/BT4/tree/tree.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>
#include <queue>
#include <functional>

using namespace std;

// =====================================================
// TREE REPRESENTATIONS
// =====================================================

// =====================================================
// 1. ARRAY OF PARENTS REPRESENTATION
// =====================================================
class ArrayOfParents {
private:
    int num_nodes;
    vector<int> parent;  // parent[v] = parent of v, or -1 if root
    int root_node;
    
public:
    ArrayOfParents(int n) : num_nodes(n), root_node(-1) {
        parent.assign(n, -1);
    }
    
    void setParent(int child, int par) {
        if (child >= 0 && child < num_nodes) {
            parent[child] = par;
        }
    }
    
    void setRoot(int r) {
        if (r >= 0 && r < num_nodes) {
            root_node = r;
            parent[r] = -1;
        }
    }
    
    int getParent(int v) const {
        if (v >= 0 && v < num_nodes) {
            return parent[v];
        }
        return -1;
    }
    
    bool isRoot(int v) const {
        return v >= 0 && v < num_nodes && parent[v] == -1;
    }
    
    int getRoot() const {
        if (root_node != -1) return root_node;
        
        // Find root by scanning array
        for (int i = 0; i < num_nodes; i++) {
            if (parent[i] == -1) {
                return i;
            }
        }
        return -1;
    }
    
    vector<int> getChildren(int v) const {
        vector<int> children;
        for (int i = 0; i < num_nodes; i++) {
            if (parent[i] == v) {
                children.push_back(i);
            }
        }
        return children;
    }
    
    void display() const {
        cout << "Array of Parents:" << endl;
        for (int i = 0; i < num_nodes; i++) {
            cout << "Node " << i << ": parent = " << parent[i] << endl;
        }
    }
    
    int getNumNodes() const { return num_nodes; }
    const vector<int>& getParentArray() const { return parent; }
};

// =====================================================
// 2. FIRST-CHILD NEXT-SIBLING REPRESENTATION
// =====================================================
class FirstChildNextSibling {
private:
    int num_nodes;
    vector<int> first_child;  // first_child[v] = first child of v, or -1 if leaf
    vector<int> next_sibling; // next_sibling[v] = next sibling of v, or -1 if last
    int root_node;
    
public:
    FirstChildNextSibling(int n) : num_nodes(n), root_node(-1) {
        first_child.assign(n, -1);
        next_sibling.assign(n, -1);
    }
    
    void setFirstChild(int parent, int child) {
        if (parent >= 0 && parent < num_nodes) {
            first_child[parent] = child;
        }
    }
    
    void setNextSibling(int current, int next) {
        if (current >= 0 && current < num_nodes) {
            next_sibling[current] = next;
        }
    }
    
    void setRoot(int r) {
        root_node = r;
    }
    
    int getFirstChild(int v) const {
        if (v >= 0 && v < num_nodes) {
            return first_child[v];
        }
        return -1;
    }
    
    int getNextSibling(int v) const {
        if (v >= 0 && v < num_nodes) {
            return next_sibling[v];
        }
        return -1;
    }
    
    bool isLeaf(int v) const {
        return v >= 0 && v < num_nodes && first_child[v] == -1;
    }
    
    bool isLastChild(int v) const {
        return v >= 0 && v < num_nodes && next_sibling[v] == -1;
    }
    
    int getRoot() const {
        return root_node;
    }
    
    vector<int> getChildren(int v) const {
        vector<int> children;
        int child = first_child[v];
        while (child != -1) {
            children.push_back(child);
            child = next_sibling[child];
        }
        return children;
    }
    
    int getParent(int v) const {
        // O(n) operation - scan to find parent
        for (int i = 0; i < num_nodes; i++) {
            int child = first_child[i];
            while (child != -1) {
                if (child == v) return i;
                child = next_sibling[child];
            }
        }
        return -1;
    }
    
    void display() const {
        cout << "First-Child Next-Sibling:" << endl;
        for (int i = 0; i < num_nodes; i++) {
            cout << "Node " << i << ": first_child = " << first_child[i] 
                 << ", next_sibling = " << next_sibling[i] << endl;
        }
    }
    
    int getNumNodes() const { return num_nodes; }
    const vector<int>& getFirstChildArray() const { return first_child; }
    const vector<int>& getNextSiblingArray() const { return next_sibling; }
};

// =====================================================
// 3. GRAPH-BASED REPRESENTATION
// =====================================================
class GraphBasedTree {
private:
    int num_nodes;
    vector<vector<int>> children;  // adjacency list for children
    vector<int> parent;            // parent[v] = parent of v, or -1 if root
    int root_node;
    
public:
    GraphBasedTree(int n) : num_nodes(n), root_node(-1) {
        children.assign(n, vector<int>());
        parent.assign(n, -1);
    }
    
    void addEdge(int par, int child) {
        if (par >= 0 && par < num_nodes && child >= 0 && child < num_nodes) {
            children[par].push_back(child);
            parent[child] = par;
        }
    }
    
    void setRoot(int r) {
        if (r >= 0 && r < num_nodes) {
            root_node = r;
            parent[r] = -1;
        }
    }
    
    int getParent(int v) const {
        if (v >= 0 && v < num_nodes) {
            return parent[v];
        }
        return -1;
    }
    
    vector<int> getChildren(int v) const {
        if (v >= 0 && v < num_nodes) {
            return children[v];
        }
        return vector<int>();
    }
    
    bool isRoot(int v) const {
        return v >= 0 && v < num_nodes && parent[v] == -1;
    }
    
    bool isLeaf(int v) const {
        return v >= 0 && v < num_nodes && children[v].empty();
    }
    
    int getRoot() const {
        if (root_node != -1) return root_node;
        
        // Find root
        for (int i = 0; i < num_nodes; i++) {
            if (parent[i] == -1) {
                return i;
            }
        }
        return -1;
    }
    
    int getFirstChild(int v) const {
        if (v >= 0 && v < num_nodes && !children[v].empty()) {
            return children[v][0];
        }
        return -1;
    }
    
    void display() const {
        cout << "Graph-Based Tree:" << endl;
        for (int i = 0; i < num_nodes; i++) {
            cout << "Node " << i << ": children = [";
            for (size_t j = 0; j < children[i].size(); j++) {
                cout << children[i][j];
                if (j < children[i].size() - 1) cout << ", ";
            }
            cout << "], parent = " << parent[i] << endl;
        }
    }
    
    int getNumNodes() const { return num_nodes; }
    const vector<vector<int>>& getChildrenList() const { return children; }
    const vector<int>& getParentArray() const { return parent; }
};


// CONVERTER 1: Array of Parents → First-Child Next-Sibling
FirstChildNextSibling arrayOfParentsToFirstChildNextSibling(const ArrayOfParents& ap) {
    int n = ap.getNumNodes();
    FirstChildNextSibling fcns(n);
    
    // Build children lists for each node
    vector<vector<int>> children_lists(n);
    for (int i = 0; i < n; i++) {
        int parent = ap.getParent(i);
        if (parent != -1) {
            children_lists[parent].push_back(i);
        }
    }
    
    // Set root
    fcns.setRoot(ap.getRoot());
    
    // Build first-child and next-sibling links
    for (int u = 0; u < n; u++) {
        if (!children_lists[u].empty()) {
            // Sort children for consistent ordering
            sort(children_lists[u].begin(), children_lists[u].end());
            
            // Set first child
            fcns.setFirstChild(u, children_lists[u][0]);
            
            // Set next-sibling links
            for (size_t i = 0; i < children_lists[u].size() - 1; i++) {
                fcns.setNextSibling(children_lists[u][i], children_lists[u][i + 1]);
            }
            // Last child's next_sibling remains -1 (already initialized)
        }
    }
    
    return fcns;
}

// CONVERTER 2: First-Child Next-Sibling → Array of Parents
void dfsSetParents(const FirstChildNextSibling& fcns, ArrayOfParents& ap, int node, int parent) {
    ap.setParent(node, parent);
    
    // Visit all children
    int child = fcns.getFirstChild(node);
    while (child != -1) {
        dfsSetParents(fcns, ap, child, node);
        child = fcns.getNextSibling(child);
    }
}

ArrayOfParents firstChildNextSiblingToArrayOfParents(const FirstChildNextSibling& fcns) {
    int n = fcns.getNumNodes();
    ArrayOfParents ap(n);
    
    // Set root
    int root = fcns.getRoot();
    if (root != -1) {
        ap.setRoot(root);
        dfsSetParents(fcns, ap, root, -1);
    }
    
    return ap;
}

// CONVERTER 3: Array of Parents → Graph-Based
GraphBasedTree arrayOfParentsToGraphBased(const ArrayOfParents& ap) {
    int n = ap.getNumNodes();
    GraphBasedTree gt(n);
    
    // Set root
    int root = ap.getRoot();
    if (root != -1) {
        gt.setRoot(root);
    }
    
    // Add edges from parent to children
    for (int i = 0; i < n; i++) {
        int parent = ap.getParent(i);
        if (parent != -1) {
            gt.addEdge(parent, i);
        }
    }
    
    return gt;
}

// CONVERTER 4: Graph-Based → Array of Parents
ArrayOfParents graphBasedToArrayOfParents(const GraphBasedTree& gt) {
    int n = gt.getNumNodes();
    ArrayOfParents ap(n);
    
    // Set root
    int root = gt.getRoot();
    if (root != -1) {
        ap.setRoot(root);
    }
    
    // Extract parent relationships
    for (int i = 0; i < n; i++) {
        int parent = gt.getParent(i);
        ap.setParent(i, parent);
    }
    
    return ap;
}

// CONVERTER 5: First-Child Next-Sibling → Graph-Based
void dfsAddEdges(const FirstChildNextSibling& fcns, GraphBasedTree& gt, int node, int parent) {
    if (parent != -1) {
        gt.addEdge(parent, node);
    }
    
    // Visit all children
    int child = fcns.getFirstChild(node);
    while (child != -1) {
        dfsAddEdges(fcns, gt, child, node);
        child = fcns.getNextSibling(child);
    }
}

GraphBasedTree firstChildNextSiblingToGraphBased(const FirstChildNextSibling& fcns) {
    int n = fcns.getNumNodes();
    GraphBasedTree gt(n);
    
    // Set root
    int root = fcns.getRoot();
    if (root != -1) {
        gt.setRoot(root);
        dfsAddEdges(fcns, gt, root, -1);
    }
    
    return gt;
}

// CONVERTER 6: Graph-Based → First-Child Next-Sibling
FirstChildNextSibling graphBasedToFirstChildNextSibling(const GraphBasedTree& gt) {
    int n = gt.getNumNodes();
    FirstChildNextSibling fcns(n);
    
    // Set root
    int root = gt.getRoot();
    if (root != -1) {
        fcns.setRoot(root);
    }
    
    // Build first-child and next-sibling links
    for (int u = 0; u < n; u++) {
        vector<int> children = gt.getChildren(u);
        
        if (!children.empty()) {
            // Sort children for consistent ordering
            sort(children.begin(), children.end());
            
            // Set first child
            fcns.setFirstChild(u, children[0]);
            
            // Set next-sibling links
            for (size_t i = 0; i < children.size() - 1; i++) {
                fcns.setNextSibling(children[i], children[i + 1]);
            }
            // Last child's next_sibling remains -1
        }
    }
    
    return fcns;
}

// =====================================================
// FILE READER FOR TREES
// =====================================================
ArrayOfParents readTreeFromFile(const string& filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cout << "Cannot open file: " << filename << endl;
        return ArrayOfParents(0);
    }
    
    string line;
    getline(file, line);
    int num_nodes = stoi(line);
    
    ArrayOfParents tree(num_nodes);
    
    while (getline(file, line)) {
        if (line.empty()) continue;
        
        istringstream iss(line);
        string type;
        iss >> type;
        
        if (type == "root") {
            int root;
            iss >> root;
            tree.setRoot(root);
            cout << "Set root: " << root << endl;
        } else if (type == "edge") {
            int parent, child;
            iss >> parent >> child;
            tree.setParent(child, parent);
            cout << "Added edge: " << parent << " -> " << child << endl;
        }
    }
    
    file.close();
    return tree;
}

// =====================================================
// TEST FUNCTION
// =====================================================
void testAllTreeConverters() {
    cout << "TEST: All 6 Tree Representation Converters" << endl;
    cout << "===========================================" << endl;
    
    // Try to read from file, otherwise create sample tree
    ArrayOfParents original = readTreeFromFile("tree_sample.inp");
    
    if (original.getNumNodes() == 0) {
        cout << "Failed to read tree from file. Using sample data instead." << endl;

        ArrayOfParents sample(7);
        sample.setRoot(0);
        sample.setParent(1, 0);
        sample.setParent(2, 0);
        sample.setParent(3, 0);
        sample.setParent(4, 1);
        sample.setParent(5, 1);
        sample.setParent(6, 2);
        original = sample;
    }
    
    cout << "\nOriginal Array of Parents:" << endl;
    original.display();
    
    // Test from Array of Parents
    cout << "\n=== 1. FROM ARRAY OF PARENTS ===" << endl;
    auto fcns_result = arrayOfParentsToFirstChildNextSibling(original);
    auto gt_result = arrayOfParentsToGraphBased(original);
    
    cout << "1.1. ArrayOfParents → FirstChildNextSibling:" << endl;
    fcns_result.display();
    cout << "1.2. ArrayOfParents → GraphBased:" << endl;
    gt_result.display();
    
    // Test from First-Child Next-Sibling
    cout << "\n=== 2. FROM FIRST-CHILD NEXT-SIBLING ===" << endl;
    cout << "2.1. FirstChildNextSibling → ArrayOfParents:" << endl;
    firstChildNextSiblingToArrayOfParents(fcns_result).display();
    cout << "2.2. FirstChildNextSibling → GraphBased:" << endl;
    firstChildNextSiblingToGraphBased(fcns_result).display();
    
    // Test from Graph-Based
    cout << "\n=== 3. FROM GRAPH-BASED ===" << endl;
    cout << "3.1. GraphBased → ArrayOfParents:" << endl;
    graphBasedToArrayOfParents(gt_result).display();
    cout << "3.2. GraphBased → FirstChildNextSibling:" << endl;
    graphBasedToFirstChildNextSibling(gt_result).display();
    
    cout << "\nDONE" << endl;
}


int main() {
    testAllTreeConverters();
    return 0;
}

File: code/Project_4/BT4/tree/tree.py
from typing import List, Optional
from collections import defaultdict

class ArrayOfParents:
    """Array of Parents representation của tree"""
    
    def __init__(self, num_nodes: int):
        self.num_nodes = num_nodes
        self.parent = [-1] * num_nodes  # parent[v] = parent của v, hoặc -1 nếu root
        self.root_node = -1
    
    def set_parent(self, child: int, parent: int):
        if 0 <= child < self.num_nodes:
            self.parent[child] = parent
    
    def set_root(self, root: int):
        if 0 <= root < self.num_nodes:
            self.root_node = root
            self.parent[root] = -1
    
    def get_parent(self, v: int) -> int:
        if 0 <= v < self.num_nodes:
            return self.parent[v]
        return -1
    
    def is_root(self, v: int) -> bool:
        return 0 <= v < self.num_nodes and self.parent[v] == -1
    
    def get_root(self) -> int:
        if self.root_node != -1:
            return self.root_node
        
        # Tìm root bằng cách scan array
        for i in range(self.num_nodes):
            if self.parent[i] == -1:
                return i
        return -1
    
    def get_children(self, v: int) -> List[int]:
        children = []
        for i in range(self.num_nodes):
            if self.parent[i] == v:
                children.append(i)
        return children
    
    def display(self):
        print("Array of Parents:")
        for i in range(self.num_nodes):
            print(f"Node {i}: parent = {self.parent[i]}")


class FirstChildNextSibling:
    """First-Child Next-Sibling representation của tree"""
    
    def __init__(self, num_nodes: int):
        self.num_nodes = num_nodes
        self.first_child = [-1] * num_nodes   
        self.next_sibling = [-1] * num_nodes 
        self.root_node = -1
    
    def set_first_child(self, parent: int, child: int):
        if 0 <= parent < self.num_nodes:
            self.first_child[parent] = child
    
    def set_next_sibling(self, current: int, next_node: int):
        if 0 <= current < self.num_nodes:
            self.next_sibling[current] = next_node
    
    def set_root(self, root: int):
        self.root_node = root
    
    def get_first_child(self, v: int) -> int:
        if 0 <= v < self.num_nodes:
            return self.first_child[v]
        return -1
    
    def get_next_sibling(self, v: int) -> int:
        if 0 <= v < self.num_nodes:
            return self.next_sibling[v]
        return -1
    
    def is_leaf(self, v: int) -> bool:
        return 0 <= v < self.num_nodes and self.first_child[v] == -1
    
    def is_last_child(self, v: int) -> bool:
        return 0 <= v < self.num_nodes and self.next_sibling[v] == -1
    
    def get_root(self) -> int:
        return self.root_node
    
    def get_children(self, v: int) -> List[int]:
        children = []
        child = self.first_child[v]
        while child != -1:
            children.append(child)
            child = self.next_sibling[child]
        return children
    
    def get_parent(self, v: int) -> int:
        """O(n) operation - scan để tìm parent"""
        for i in range(self.num_nodes):
            child = self.first_child[i]
            while child != -1:
                if child == v:
                    return i
                child = self.next_sibling[child]
        return -1
    
    def display(self):
        print("First-Child Next-Sibling:")
        for i in range(self.num_nodes):
            print(f"Node {i}: first_child = {self.first_child[i]}, next_sibling = {self.next_sibling[i]}")


class GraphBasedTree:
    """Graph-Based representation của tree"""
    
    def __init__(self, num_nodes: int):
        self.num_nodes = num_nodes
        self.children = [[] for _ in range(num_nodes)]  # adjacency list cho children
        self.parent = [-1] * num_nodes                  # parent[v] = parent của v, hoặc -1 nếu root
        self.root_node = -1
    
    def add_edge(self, parent: int, child: int):
        if 0 <= parent < self.num_nodes and 0 <= child < self.num_nodes:
            self.children[parent].append(child)
            self.parent[child] = parent
    
    def set_root(self, root: int):
        if 0 <= root < self.num_nodes:
            self.root_node = root
            self.parent[root] = -1
    
    def get_parent(self, v: int) -> int:
        if 0 <= v < self.num_nodes:
            return self.parent[v]
        return -1
    
    def get_children(self, v: int) -> List[int]:
        if 0 <= v < self.num_nodes:
            return self.children[v].copy()
        return []
    
    def is_root(self, v: int) -> bool:
        return 0 <= v < self.num_nodes and self.parent[v] == -1
    
    def is_leaf(self, v: int) -> bool:
        return 0 <= v < self.num_nodes and len(self.children[v]) == 0
    
    def get_root(self) -> int:
        if self.root_node != -1:
            return self.root_node
        
        # Tìm root
        for i in range(self.num_nodes):
            if self.parent[i] == -1:
                return i
        return -1
    
    def get_first_child(self, v: int) -> int:
        if 0 <= v < self.num_nodes and len(self.children[v]) > 0:
            return self.children[v][0]
        return -1
    
    def display(self):
        print("Graph-Based Tree:")
        for i in range(self.num_nodes):
            print(f"Node {i}: children = {self.children[i]}, parent = {self.parent[i]}")


# =====================================================
# 6 TREE CONVERTERS
# =====================================================

def array_of_parents_to_first_child_next_sibling(ap: ArrayOfParents) -> FirstChildNextSibling:
    """CONVERTER 1: Array of Parents → First-Child Next-Sibling"""
    n = ap.num_nodes
    fcns = FirstChildNextSibling(n)
    
    # Build children lists cho mỗi node
    children_lists = [[] for _ in range(n)]
    for i in range(n):
        parent = ap.get_parent(i)
        if parent != -1:
            children_lists[parent].append(i)
    
    # Set root
    fcns.set_root(ap.get_root())
    
    # Build first-child và next-sibling links
    for u in range(n):
        if children_lists[u]:
            # Sort children cho consistent ordering
            children_lists[u].sort()
            
            # Set first child
            fcns.set_first_child(u, children_lists[u][0])
            
            # Set next-sibling links
            for i in range(len(children_lists[u]) - 1):
                fcns.set_next_sibling(children_lists[u][i], children_lists[u][i + 1])
            # Last child's next_sibling remains -1
    
    return fcns


def first_child_next_sibling_to_array_of_parents(fcns: FirstChildNextSibling) -> ArrayOfParents:
    """CONVERTER 2: First-Child Next-Sibling → Array of Parents"""
    n = fcns.num_nodes
    ap = ArrayOfParents(n)
    
    # Set root
    root = fcns.get_root()
    if root != -1:
        ap.set_root(root)
    
    # DFS để set parent relationships
    def dfs(node: int, parent: int):
        ap.set_parent(node, parent)
        
        # Visit all children
        child = fcns.get_first_child(node)
        while child != -1:
            dfs(child, node)
            child = fcns.get_next_sibling(child)
    
    if root != -1:
        dfs(root, -1)
    
    return ap


def array_of_parents_to_graph_based(ap: ArrayOfParents) -> GraphBasedTree:
    """CONVERTER 3: Array of Parents → Graph-Based"""
    n = ap.num_nodes
    gt = GraphBasedTree(n)
    
    # Set root
    root = ap.get_root()
    if root != -1:
        gt.set_root(root)
    
    # Add edges từ parent đến children
    for i in range(n):
        parent = ap.get_parent(i)
        if parent != -1:
            gt.add_edge(parent, i)
    
    return gt


def graph_based_to_array_of_parents(gt: GraphBasedTree) -> ArrayOfParents:
    """CONVERTER 4: Graph-Based → Array of Parents"""
    n = gt.num_nodes
    ap = ArrayOfParents(n)
    
    # Set root
    root = gt.get_root()
    if root != -1:
        ap.set_root(root)
    
    # Extract parent relationships
    for i in range(n):
        parent = gt.get_parent(i)
        ap.set_parent(i, parent)
    
    return ap


def first_child_next_sibling_to_graph_based(fcns: FirstChildNextSibling) -> GraphBasedTree:
    """CONVERTER 5: First-Child Next-Sibling → Graph-Based"""
    n = fcns.num_nodes
    gt = GraphBasedTree(n)
    
    # Set root
    root = fcns.get_root()
    if root != -1:
        gt.set_root(root)
    
    # DFS để build edges
    def dfs(node: int, parent: int):
        if parent != -1:
            gt.add_edge(parent, node)
        
        # Visit all children
        child = fcns.get_first_child(node)
        while child != -1:
            dfs(child, node)
            child = fcns.get_next_sibling(child)
    
    if root != -1:
        dfs(root, -1)
    
    return gt


def graph_based_to_first_child_next_sibling(gt: GraphBasedTree) -> FirstChildNextSibling:
    """CONVERTER 6: Graph-Based → First-Child Next-Sibling"""
    n = gt.num_nodes
    fcns = FirstChildNextSibling(n)
    
    # Set root
    root = gt.get_root()
    if root != -1:
        fcns.set_root(root)
    
    # Build first-child và next-sibling links
    for u in range(n):
        children = gt.get_children(u)
        
        if children:
            # Sort children cho consistent ordering
            children.sort()
            
            # Set first child
            fcns.set_first_child(u, children[0])
            
            # Set next-sibling links
            for i in range(len(children) - 1):
                fcns.set_next_sibling(children[i], children[i + 1])
            # Last child's next_sibling remains -1
    
    return fcns


# =====================================================
# FILE READER FOR TREES
# =====================================================
def read_tree_from_file(filename: str) -> ArrayOfParents:
    try:
        with open(filename, 'r') as file:
            lines = file.readlines()
        
        num_nodes = int(lines[0].strip())
        tree = ArrayOfParents(num_nodes)
        
        for i in range(1, len(lines)):
            line = lines[i].strip()
            if not line:
                continue
            
            parts = line.split()
            if len(parts) >= 2:
                if parts[0] == "root":
                    root = int(parts[1])
                    tree.set_root(root)
                    print(f"Set root: {root}")
                elif parts[0] == "edge" and len(parts) >= 3:
                    parent = int(parts[1])
                    child = int(parts[2])
                    tree.set_parent(child, parent)
                    print(f"Added edge: {parent} -> {child}")
        
        return tree
        
    except FileNotFoundError:
        print(f"Cannot open file: {filename}")
        return ArrayOfParents(0)
    except Exception as e:
        print(f"Error reading file: {e}")
        return ArrayOfParents(0)


def test_all_tree_converters():
    print("TEST: All 6 Tree Representation Converters")
    print("===========================================")
    
    original = read_tree_from_file("..\tree_sample.inp")

    original.display()
    
    # Test from Array of Parents
    print("\n=== 1. FROM ARRAY OF PARENTS ===")
    fcns_result = array_of_parents_to_first_child_next_sibling(original)
    gt_result = array_of_parents_to_graph_based(original)
    
    print("1.1. ArrayOfParents → FirstChildNextSibling:")
    fcns_result.display()
    print("1.2. ArrayOfParents → GraphBased:")
    gt_result.display()
    
    # Test from First-Child Next-Sibling
    print("\n=== 2. FROM FIRST-CHILD NEXT-SIBLING ===")
    print("2.1. FirstChildNextSibling → ArrayOfParents:")
    first_child_next_sibling_to_array_of_parents(fcns_result).display()
    print("2.2. FirstChildNextSibling → GraphBased:")
    first_child_next_sibling_to_graph_based(fcns_result).display()
    
    # Test from Graph-Based
    print("\n=== 3. FROM GRAPH-BASED ===")
    print("3.1. GraphBased → ArrayOfParents:")
    graph_based_to_array_of_parents(gt_result).display()
    print("3.2. GraphBased → FirstChildNextSibling:")
    graph_based_to_first_child_next_sibling(gt_result).display()
    
    print("\nDONE")


if __name__ == "__main__":
    test_all_tree_converters()